Using Watcher<Event> leads to "JsonMappingException: No resource type found for:v1#Event"

When I try to listen for events with a `Watcher<Event>` like this...

```java
client.v1().events().inNamespace(namespaceName).watch(watcher)
```

... then the watcher doesn't receive any events as expected, and I see stack traces like this instead:

```
WARN [2020-07-02 13:25:40,217] io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager: Exec Failure
io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
        at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:64)
        at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:53)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:248)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:167)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:152)
        at io.fabric8.kubernetes.client.dsl.internal.WatchHTTPManager.readWatchEvent(WatchHTTPManager.java:292)
        at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onMessage(WatchConnectionManager.java:230)
        at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:323)
        at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:274)
        at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:214)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.java:203)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: No resource type found for:v1#Event
 at [Source: (BufferedInputStream); line: 1, column: 970] (through reference chain: io.fabric8.kubernetes.api.model.WatchEvent["object"])
        at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:274)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer.fromObjectNode(KubernetesDeserializer.java:85)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:57)
        at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:46)
        at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4218)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3259)
        at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:246)
        ... 14 common frames omitted
```

I started seeing this in an internal project after upgrading kubernetes-client from 4.9.0 to 4.10.2.

I'm working around the issue in my project by registering a `KubernetesResourceMappingProvider` that maps `"v1#Event"` to `io.fabric8.kubernetes.api.model.Event.class`.

I can't share that internal project, but I created a standalone project that demonstrates the same error along with a workaround.  See: https://github.com/michaelhixson/kubernetes-client-event-test

My dev environment where I'm seeing this issue:
* Windows 10
* Java 14.0.1
* Maven 3.6.3
* Docker Desktop 2.3.0.3
* Docker: 19.03.8
* Kubernetes: v1.16.5
* kubernetes-client library: 4.10.2

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 * Fix Raw CustomResource API path generation to not having trailing slash
 * Fix #2131: Failing to parse CustomResourceDefinition with OpenAPIV3Schema using JSONSchemaPropOr\* fields
 * Fix #2297: Resuscitate ProjectRequestHandler in openshift-client
+* Fix #2328: Failure in deserialization while watching events
 * Fix KubernetesAttributesExctractor to extract metadata from unregistered custom resources, such when using Raw CustomResource API 
 * Fix #2296: No adapter available for type:interface io.fabric8.kubernetes.client.dsl.V1APIGroupDSL
 

--- a/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
+++ b/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java
@@ -264,10 +264,11 @@ public class KubernetesDeserializer extends JsonDeserializer<KubernetesResource>
           String apiVersion = Helper.getAnnotationValue(clazz, ApiVersion.class);
           if (apiGroup != null && !apiGroup.isEmpty() && apiVersion != null && !apiVersion.isEmpty()) {
             return createKey(apiGroup + "/" + apiVersion, clazz.getSimpleName());
+          } else if (apiVersion != null && !apiVersion.isEmpty()) {
+            return createKey(apiVersion, clazz.getSimpleName());
           }
           return clazz.getSimpleName();
         }
-
 
         private Class<? extends KubernetesResource> loadClassIfExists(String className) {
             try {

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/EventTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/EventTest.java
@@ -1,0 +1,77 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes.client.mock;
+
+import io.fabric8.kubernetes.api.model.Event;
+import io.fabric8.kubernetes.api.model.EventBuilder;
+import io.fabric8.kubernetes.api.model.WatchEvent;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.server.mock.KubernetesServer;
+import org.junit.Rule;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@EnableRuleMigrationSupport
+class EventTest {
+  @Rule
+  public KubernetesServer server = new KubernetesServer();
+
+  @Test
+  @DisplayName("Should be able to watch events in specified namespace")
+  void watch() throws InterruptedException {
+    // Given
+    Event testEvent = new EventBuilder()
+      .withNewMetadata()
+      .withName("nginx-deployment-54f57cf6bf-84ssh.161e26b7b76b629e")
+      .withNamespace("ns1")
+      .endMetadata()
+      .build();
+
+    server.expect()
+      .withPath("/api/v1/namespaces/ns1/events?watch=true")
+      .andUpgradeToWebSocket().open().waitFor(50)
+      .andEmit(new WatchEvent(testEvent, "ADDED"))
+      .done().once();
+    KubernetesClient client = server.getClient();
+    final CountDownLatch eventReceivedLatch = new CountDownLatch(1);
+
+    // When
+    Watch watch = client.v1().events().inNamespace("ns1").watch(new Watcher<Event>() {
+      @Override
+      public void eventReceived(Action action, Event resource) {
+        eventReceivedLatch.countDown();
+      }
+
+      @Override
+      public void onClose(KubernetesClientException cause) {
+
+      }
+    });
+
+    // Then
+    assertTrue(eventReceivedLatch.await(1, TimeUnit.SECONDS));
+    watch.close();
+  }
+}

